[Swift in WebKit] Fix the modularization of WebGPU

rr-codes opened the pull request:
#### c3f6cc3af29cfe85fd955830d1a2b59da4cc670e
<pre>
[Swift in WebKit] Fix the modularization of WebGPU
<a href="https://bugs.webkit.org/show_bug.cgi?id=304043">https://bugs.webkit.org/show_bug.cgi?id=304043</a>
<a href="https://rdar.apple.com/166342376">rdar://166342376</a>

Reviewed by Mike Wyrzykowski.

Make some changes to correct all modularization issues in WebGPU.

Build configuration stuff:

- Enable the module verifier in Objective C++ mode for WebGPU.
- Set SWIFT_LIBRARY_LEVEL to its proper value.
- The enablement of `ENABLE_WEBGPU_SWIFT` for macOS was conditionalized on the architecture.
However, in some parts of the build, this value is `undefined`. This leads to incredibly
confusing and surprising results. To address this, and to allow module verification to
actually be conditionalized on `ENABLE_WEBGPU_SWIFT`, just enable `ENABLE_WEBGPU_SWIFT` at
build time on all architectures on macOS.
- Set the `-emit-clang-header-min-access` flag to `internal` so that Swift declarations need
not be `public`; not only is this just reasonably better, it&apos;s also necessary for reasons
described below.

Access level / exposed interface stuff:

- Previously, there were public Swift functions that depended on project-level C++ types.
This is an access level violation and modules violation and probably some other violations,
and was leading to a variety of build issues. By making all the Swift symbols internal
instead of public, this fixes that.

- Add more explicit access level specifiers to the import statements for clarity.

- Some functions and types were declared in the public module, and then implemented in the
internal project module. This is a modules violation and was causing issues with Cxx
interoperability being able to find the necessary functions and types. Fix this by just
moving the implementation to the public module.

Modules stuff:

- Add missing header includes and module imports
- Add a private module map, `WebGPU_Private.modulemap` for the private headers in WebGPU
- Fix the existing public module map
- Fix the existing internal module map; turns out, it&apos;s invalid and not supported for a
framework project internal module to have an umbrella header, since those are meant to
comprise everything in a single directory, which isn&apos;t the case for the structure of project
internal headers; it essentially was only &quot;working&quot; by happenstance. Consequently, fix this
by explicitly creating all the sub-modules, and make them explicit for clarity and to also
avoid any confusing implicit imports.

Misc. stuff:

- Fix compile time errors now that x86_64 compiles the Swift code
- Device.h needs to out-of-line the implementation of `vertexBufferIndexForBindGroup` since
WebGPU cannot have a header file dependent on WGSL because WGSL is not a well-formed library/
module.

* Source/WTF/Configurations/WTF.xcconfig:
* Source/WebGPU/Configurations/WebGPU.xcconfig:
* Source/WebGPU/WebGPU.xcodeproj/project.pbxproj:
* Source/WebGPU/WebGPU/Buffer.h:
* Source/WebGPU/WebGPU/Buffer.mm:
* Source/WebGPU/WebGPU/Buffer.swift:
(Buffer_copyFrom_thunk(_:from:offset:)):
(Buffer_getMappedRange_thunk(_:offset:size:)):
* Source/WebGPU/WebGPU/CommandBuffer.h:
* Source/WebGPU/WebGPU/CommandEncoder.h:
* Source/WebGPU/WebGPU/CommandEncoder.mm:
* Source/WebGPU/WebGPU/CommandEncoder.swift:
(WebGPU.finish(_:)):
(WebGPU.clearTextureIfNeeded(_:slice:)):
(resolveQuerySet(_:querySet:firstQuery:queryCount:destination:destinationOffset:)): Deleted.
(CommandEncoder_copyBufferToBuffer_thunk(_:source:sourceOffset:destination:destinationOffset:size:)): Deleted.
(WebGPU.clearTextureIfNeeded(_:slice:device:blitCommandEncoder:)): Deleted.
* Source/WebGPU/WebGPU/ComputePassEncoder.h:
* Source/WebGPU/WebGPU/DDMesh.mm:
* Source/WebGPU/WebGPU/Device.h:
(WebGPU::Device::vertexBufferIndexForBindGroup const): Deleted.
* Source/WebGPU/WebGPU/Device.mm:
* Source/WebGPU/WebGPU/HardwareCapabilities.h:
* Source/WebGPU/WebGPU/Instance.mm:
* Source/WebGPU/WebGPU/Internal/module.modulemap:
* Source/WebGPU/WebGPU/QuerySet.h:
* Source/WebGPU/WebGPU/Queue.mm:
* Source/WebGPU/WebGPU/Queue.swift:
(Queue_writeBuffer_thunk(_:buffer:bufferOffset:data:)):
(WebGPU.writeBuffer(_:bufferOffset:data:)):
* Source/WebGPU/WebGPU/StdLibExtras.swift:
* Source/WebGPU/WebGPU/Texture.h:
* Source/WebGPU/WebGPU/TextureView.h:
* Source/WebGPU/WebGPU/USDModel.swift:
* Source/WebGPU/WebGPU/WGPUBufferImpl.h: Copied from Source/WebGPU/WebGPU/Internal/WebGPUSwiftInternal.h.
* Source/WebGPU/WebGPU/WGPUQuerySetImpl.h: Copied from Source/WebGPU/WebGPU/Internal/WebGPUSwiftInternal.h.
* Source/WebGPU/WebGPU/WGPUTextureImpl.h: Copied from Source/WebGPU/WebGPU/Internal/WebGPUSwiftInternal.h.
* Source/WebGPU/WebGPU/WGPUTextureViewImpl.h: Renamed from Source/WebGPU/WebGPU/Internal/WebGPUSwiftInternal.h.
* Source/WebGPU/WebGPU/WebGPU.h:
* Source/WebGPU/WebGPU/WebGPU.modulemap:
* Source/WebGPU/WebGPU/WebGPU_Private.modulemap: Added.

Canonical link: <a href="https://commits.webkit.org/304524@main">https://commits.webkit.org/304524@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/27a6de05a67d19ce6f9063d78bd125fd6987cce9

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/135778 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/8154 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/47071 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/143492 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/87402 "Built successfully") | ⏳ 🛠 ios-apple 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/8812 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/8001 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/103764 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/71696 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/a0762dfc-9a69-4962-b3f5-ac1b1153f4e1) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/138724 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/162/builds/6353 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/121694 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/84642 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/6f3896b3-6dab-4485-a1ca-6a62010f1ce5) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/154/builds/6099 "Passed tests") | [✅ 🧪 api-mac-debug](https://ews-build.webkit.org/#/builders/165/builds/3718 "Passed tests") | [✅ 🛠 wpe-cairo-libwebrtc](https://ews-build.webkit.org/#/builders/166/builds/4098 "Built successfully") | | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/127760 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/115330 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/39883 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/146241 "Built successfully") | | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/134261 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/7842 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/40450 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/112125 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/7868 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/6578 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/112506 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/28561 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/5975 "Passed tests") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/117995 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/61771 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/7888 "Built successfully") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/36100 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/167082 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/7622 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/71438 "Built successfully") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/43617 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/7849 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/7717 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

mwyrzykowski reviewed (approved):
looks great thank you :)

rr-codes committed:
[Swift in WebKit] Fix the modularization of WebGPU
https://bugs.webkit.org/show_bug.cgi?id=304043
rdar://166342376

Reviewed by Mike Wyrzykowski.

Make some changes to correct all modularization issues in WebGPU.

Build configuration stuff:

- Enable the module verifier in Objective C++ mode for WebGPU.
- Set SWIFT_LIBRARY_LEVEL to its proper value.
- The enablement of `ENABLE_WEBGPU_SWIFT` for macOS was conditionalized on the architecture.
However, in some parts of the build, this value is `undefined`. This leads to incredibly
confusing and surprising results. To address this, and to allow module verification to
actually be conditionalized on `ENABLE_WEBGPU_SWIFT`, just enable `ENABLE_WEBGPU_SWIFT` at
build time on all architectures on macOS.
- Set the `-emit-clang-header-min-access` flag to `internal` so that Swift declarations need
not be `public`; not only is this just reasonably better, it's also necessary for reasons
described below.

Access level / exposed interface stuff:

- Previously, there were public Swift functions that depended on project-level C++ types.
This is an access level violation and modules violation and probably some other violations,
and was leading to a variety of build issues. By making all the Swift symbols internal
instead of public, this fixes that.

- Add more explicit access level specifiers to the import statements for clarity.

- Some functions and types were declared in the public module, and then implemented in the
internal project module. This is a modules violation and was causing issues with Cxx
interoperability being able to find the necessary functions and types. Fix this by just
moving the implementation to the public module.

Modules stuff:

- Add missing header includes and module imports
- Add a private module map, `WebGPU_Private.modulemap` for the private headers in WebGPU
- Fix the existing public module map
- Fix the existing internal module map; turns out, it's invalid and not supported for a
framework project internal module to have an umbrella header, since those are meant to
comprise everything in a single directory, which isn't the case for the structure of project
internal headers; it essentially was only "working" by happenstance. Consequently, fix this
by explicitly creating all the sub-modules, and make them explicit for clarity and to also
avoid any confusing implicit imports.

Misc. stuff:

- Fix compile time errors now that x86_64 compiles the Swift code
- Device.h needs to out-of-line the implementation of `vertexBufferIndexForBindGroup` since
WebGPU cannot have a header file dependent on WGSL because WGSL is not a well-formed library/
module.

* Source/WTF/Configurations/WTF.xcconfig:
* Source/WebGPU/Configurations/WebGPU.xcconfig:
* Source/WebGPU/WebGPU.xcodeproj/project.pbxproj:
* Source/WebGPU/WebGPU/Buffer.h:
* Source/WebGPU/WebGPU/Buffer.mm:
* Source/WebGPU/WebGPU/Buffer.swift:
(Buffer_copyFrom_thunk(_:from:offset:)):
(Buffer_getMappedRange_thunk(_:offset:size:)):
* Source/WebGPU/WebGPU/CommandBuffer.h:
* Source/WebGPU/WebGPU/CommandEncoder.h:
* Source/WebGPU/WebGPU/CommandEncoder.mm:
* Source/WebGPU/WebGPU/CommandEncoder.swift:
(WebGPU.finish(_:)):
(WebGPU.clearTextureIfNeeded(_:slice:)):
(resolveQuerySet(_:querySet:firstQuery:queryCount:destination:destinationOffset:)): Deleted.
(CommandEncoder_copyBufferToBuffer_thunk(_:source:sourceOffset:destination:destinationOffset:size:)): Deleted.
(WebGPU.clearTextureIfNeeded(_:slice:device:blitCommandEncoder:)): Deleted.
* Source/WebGPU/WebGPU/ComputePassEncoder.h:
* Source/WebGPU/WebGPU/DDMesh.mm:
* Source/WebGPU/WebGPU/Device.h:
(WebGPU::Device::vertexBufferIndexForBindGroup const): Deleted.
* Source/WebGPU/WebGPU/Device.mm:
* Source/WebGPU/WebGPU/HardwareCapabilities.h:
* Source/WebGPU/WebGPU/Instance.mm:
* Source/WebGPU/WebGPU/Internal/module.modulemap:
* Source/WebGPU/WebGPU/QuerySet.h:
* Source/WebGPU/WebGPU/Queue.mm:
* Source/WebGPU/WebGPU/Queue.swift:
(Queue_writeBuffer_thunk(_:buffer:bufferOffset:data:)):
(WebGPU.writeBuffer(_:bufferOffset:data:)):
* Source/WebGPU/WebGPU/StdLibExtras.swift:
* Source/WebGPU/WebGPU/Texture.h:
* Source/WebGPU/WebGPU/TextureView.h:
* Source/WebGPU/WebGPU/USDModel.swift:
* Source/WebGPU/WebGPU/WGPUBufferImpl.h: Copied from Source/WebGPU/WebGPU/Internal/WebGPUSwiftInternal.h.
* Source/WebGPU/WebGPU/WGPUQuerySetImpl.h: Copied from Source/WebGPU/WebGPU/Internal/WebGPUSwiftInternal.h.
* Source/WebGPU/WebGPU/WGPUTextureImpl.h: Copied from Source/WebGPU/WebGPU/Internal/WebGPUSwiftInternal.h.
* Source/WebGPU/WebGPU/WGPUTextureViewImpl.h: Renamed from Source/WebGPU/WebGPU/Internal/WebGPUSwiftInternal.h.
* Source/WebGPU/WebGPU/WebGPU.h:
* Source/WebGPU/WebGPU/WebGPU.modulemap:
* Source/WebGPU/WebGPU/WebGPU_Private.modulemap: Added.

Canonical link: https://commits.webkit.org/304524@main
